arangod/Utils/CollectionNameResolver.cpp

```diff
@@ -319,36 +289,87 @@ std::string CollectionNameResolver::localNameLookup(TRI_voc_cid_t cid) const {
 std::shared_ptr<LogicalDataSource> CollectionNameResolver::getDataSource(
     TRI_voc_cid_t id
 ) const noexcept {
+  auto itr = _dataSourceById.find(id);
+
+  if (itr != _dataSourceById.end()) {
+    return itr->second;
+  }
+
+  std::shared_ptr<LogicalDataSource> ptr;
+
   // db server / standalone
-  if (!ServerState::isRunningInCluster(_serverRole)
-      || ServerState::isDBServer(_serverRole)) {
-    return _vocbase ? _vocbase->lookupDataSource(id) : nullptr;
+  if (!ServerState::isCoordinator(_serverRole)) {
+    ptr = _vocbase ? _vocbase->lookupDataSource(id) : nullptr;
+  } else {
+    ptr = getDataSource(getCollectionName(id));
   }
 
-  // cluster coordinator
-  auto* ci = ClusterInfo::instance();
+  if (ptr) {
+    _dataSourceById.emplace(id, ptr);
+  }
+
+  return ptr;
+}
 
-  if (!ci) {
-    return nullptr;
+std::shared_ptr<LogicalDataSource> CollectionNameResolver::getDataSource(
+  std::string const& nameOrId
+) const noexcept {
+  auto itr = _dataSourceByName.find(nameOrId);
+
+  if (itr != _dataSourceByName.end()) {
+    return itr->second;
   }
 
-  try {
-    auto name = std::to_string(id);
-    auto cinfo = ci->getCollection(_vocbase->name(), name);
+  std::shared_ptr<LogicalDataSource> ptr;
 
-    if (cinfo) {
-      return std::static_pointer_cast<LogicalDataSource>(cinfo);
+  // db server / standalone
+  if (!ServerState::isCoordinator(_serverRole)) {
+    ptr = _vocbase ? _vocbase->lookupDataSource(nameOrId) : nullptr;
+  } else if (nameOrId[0] >= '0' && nameOrId[0] <= '9') {
+    ptr = getDataSource(getCollectionName(NumberUtils::atoi_zero<TRI_voc_cid_t>(
+      nameOrId.data(), nameOrId.data() + nameOrId.size()
+    )));
+  } else {
+    // cluster coordinator
+    auto* ci = ClusterInfo::instance();
+
+    if (!ci) {
+      return nullptr;
     }
 
-    return std::static_pointer_cast<LogicalDataSource>(
-      ci->getView(_vocbase->name(), name)
-    );
-  } catch (...) {
-    LOG_TOPIC(ERR, arangodb::Logger::FIXME)
-      << "caught exception while resolving cluster data-source id: " << id;
+    try {
+      ptr = std::static_pointer_cast<LogicalDataSource>(
+        ci->getCollection(_vocbase->name(), nameOrId)
+      );
+
+      if (!ptr) {
+        ptr = std::static_pointer_cast<LogicalDataSource>(
+          ci->getView(_vocbase->name(), nameOrId)
+        );
+      }
+    } catch (...) {
+      LOG_TOPIC(ERR, arangodb::Logger::FIXME)
+        << "caught exception while resolving cluster data-source: " << nameOrId;
+    }
+  }
+
+  if (ptr) {
+    _dataSourceByName.emplace(nameOrId, ptr);
   }
 
-  return nullptr;
+  return ptr;
+}
+
+std::shared_ptr<LogicalView> CollectionNameResolver::getView(
+  TRI_voc_cid_t id
+) const noexcept {
+  return std::dynamic_pointer_cast<LogicalView>(getDataSource(id));
```

Comment:
why `dynamic_pointer_cast`?

Reply:
because getDataSource(id) returns a std::shared_pointer which can actually be a LogicalCollection or a LogicalView

Reply:
please use something like:
```
auto dataSource = getDataSource(id);
return dataSource.category() == LogicalView::category() ? std::static_pointer_cast<LogicalView>(dataSource) : nullptr;
```
instead

Reply:
done